### DISCUSSION TITLE
Struggling with 4 Wire 54154

### DISCUSSION THREAD

**Discussion by Fox49** [visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA#763] [Q&A] [answered]
Hi, I have the 54154 pump with 4 wires.

I ordered pcb v2 and got the transistor based LLC, everything is soldered and the board is on, as well as I have full access to the web interface (and control it via HA & MQTT)

My problem is that whatever I do nothing happens. The display comes on and the buttons on it light up when i press it, but nothing works. Same thing for the web interface, wether I enable God mode or not.

Here is the CIO connection and the pinout in the interface, as well as my board connected...

Can anyone help me to figure out what's wrong plz ? 

TY,

![IMG_20240717_175355](https://github.com/user-attachments/assets/8deb7271-cd3f-499a-a5ab-a591069624da)
![IMG_20240717_175321](https://github.com/user-attachments/assets/0ca687ee-42b8-4591-86ba-9db8619a9a38)
![IMG_20240717_132201](https://github.com/user-attachments/assets/ca078f7e-3d43-41c5-843f-c016feee9e9f)
![Lay-Z-Spa Module - Google Chrome](https://github.com/user-attachments/assets/aa108ead-c2d9-4452-9a0d-6ee12a365936)
![Lay-Z-Spa Module  Hardware Config - Google Chrome](https://github.com/user-attachments/assets/21db6664-2cb2-48c9-8d85-e3498dd99f33)
![IMG_20240717_175546](https://github.com/user-attachments/assets/2f73d69b-5e6a-4681-a887-1bb2015c0682)

**Comment by visualapproach** [comment]
Hi, 
1 plz post a picture of the other side of PCB (the soldering side)
2 Use v4.4.2

**Reply by Fox49** [reply]
![IMG_20240717_181538](https://github.com/user-attachments/assets/56364d1c-1b40-4afb-986f-7996438a97ea)
Thanks for the quick reply !

Just upgraded to 4.4.2, exact same behavior :-/ 

I don't know if this can be a clue, but when I toggle the "take control" button, the "pump" switch goes on and off multiple times..?

**Reply by Fox49** [reply]
https://github.com/user-attachments/assets/12ca78fe-555e-47b5-b841-b7d42bbe0c0a

**Reply by visualapproach** [reply]
The soldering looks decent except for the cables. Maybe touch them up a bit.
And you can do the hwtest to see if they are good. See manual.

**Reply by Fox49** [reply]
Yeah I thought the same while looking at it, just resoldered them nicely. 

Here are the results of the tests, everything seems ok, I don't understand.. :-/

![layzspa localhwtest - Google Chrome](https://github.com/user-attachments/assets/461260c0-df6c-4275-9832-4ca31cb86912)

**Comment by visualapproach** [comment]
click here-> [layzspa.local/debug-on/](http://layzspa.local/debug-on/)
let it run for several minutes, then click here [layzspa.local/debug-off/](http://layzspa.local/debug-off/)
Then post this file -> [log.txt](http://layzspa.local/log.txt)

**Reply by Fox49** [reply]
[log.txt](https://github.com/user-attachments/files/16272144/log.txt)

**Reply by visualapproach** [reply]
Good. Comms to/from the CIO (pump side) is not working. Check that you wired cio tx to D5 and cio rx to D2. You can just switch place and test if you want. Or check like this- the wire from dsp appears to be right. So the wire that goes to dsp rx (D3) should be the same that goes to cio tx. If you connect the dsp directly to the cio you see what wire it is.

If everything is correct you have a few options.
1- use the B ports with caps and resistors.
2- switch to other pins. There is a thread about that where someone succeeded by doing that. I'll see if I can find it.
3- all of the above

**Comment by Fox49** [comment]
OK thanks, I re-checked my wiring, and rechecked it again, and again..(lol) and it's definitely in the right position.

I soldered the resistors and wires to H1B/2B, but I don't have the capacitors (so far it's still not working but I guess it's because I miss the capacitors).

If you find me the thread that talks about switching pins.. It'll give me some reading while I wait for the capacitors to be delivered ! XD

**Reply by visualapproach** [reply]
#729

**Reply by Fox49** [reply]
He has the exact same problem as me.

**Reply by Fox49** [reply]
@visualapproach 
Do we agree that the DSP side of the board is the wiring coming FROM the DSP, and the CIO one is the same ? 

It is not the side marked DSP where I should connect the CIO (because it's TO the display) ?

Just to be sure...

**Reply by visualapproach** [reply]
Dsp goes to dsp
Cio to cio

**Comment by Fox49** [comment]
@piltous

**Reply by piltous** [reply]
Dsp seems ok.
So try different pins for cio to get messages but think that you need the full communication between cio and dsp so:
Tx cio > esp > rx dsp
Tx dsp > esp > rx cio
Thing also about the behavior expected as I mentionned previously.
If I said something wrong, the boss will correct me ;)

**Reply by Fox49** [reply]
So if I resume : 

- First things first, the main goal is that the ESP can control the CIO. So DSP can be set aside as long as everything isn't perfect between those two to begin with, right ? (But I do need it at least to power on the unit, right ?)

For ESP to control the CIO, I need to focus on the TX and RX Pins, which I think is simple : the esp send (TX) the payload with the command associated, on a GPIO port of the nodemcu, which is apparently not the default one set on the config file for my CIO, and the CIO responds to ESP on RX port to acknowledge command
-> Is this logic OK or am I misunderstanding smthg ?

@visualapproach are the default gpio pins on the nodemcu defined arbitrarily in the code or is there a logic taht I can rely on to find out what are the good ones for me ? 

- @piltous another thing that I don't understand, did you figure out what was the "04" displayed on the screen by default ?

**Reply by visualapproach** [reply]
Data flows
CIO TX -> ESP -> DSP RX
DSP TX -> ESP -> CIO RX
The pins are chosen to be the best suited, but apparently for piltous he had better luck with other pins. You can set "custom PCB" and try other pins than default. Remember to restart the ESP for the change to take effect.

**Comment by piltous** [comment]
@Fox49 yes 04 by default it seems. I still have this when I start but 2s after it switch to the correct value.

Test different pins for the one you re the less confidant and use debug feature to check the message until you have 100% for both dsg and cio

**Reply by piltous** [reply]
@Fox49  also search in the closed issues if someone has the same pump and setup than you

**Comment by Fox49** [comment]
@piltous can you post a photo of your LLC Transistor plz ? 

(And your whole setup btw)

**Comment by piltous** [comment]
![20240718_210012](https://github.com/user-attachments/assets/59af820c-5cad-43e2-9a12-6beaead9c62f)

**Reply by piltous** [reply]
Here is the mess :)

**Reply by Fox49** [reply]
TY

I don't understand how it can work when the pins are switched... 'cause in the schematics the gpio pin goes to only one port of the LLC which goes straight forward to the port in front of it :-/

![image](https://github.com/user-attachments/assets/27715690-c765-47af-b9f5-47f7174912da)

![image](https://github.com/user-attachments/assets/6b41911c-9565-4b38-b136-e8745f85d60b)

That's weird !

**Comment by Fox49** [comment]
oh sorry I understand now, you soldered the cable according to the pin change.

I see.

**Reply by piltous** [reply]
Right and no real explaination for this. I think there is a lot a variables in the loop.

**Comment by Fox49** [comment, marked as answer]
IT WORKS !!!

Drum roll.... 
..
.

Turns out I had a faulty LLC... I think it's because of the flux I used... It was conducting electricity... So I likely fried it... Or maybe the first time when I had not plugged 5V and GND on the good pins lol

Here is my final configuration for those interested in (but I think I could revert to default ports it would still work) :

![image](https://github.com/user-attachments/assets/98cf9d8f-b1c4-4d8e-b392-b3bb8aa3837b)

**Reply by piltous** [reply]
Congrats! :)

**Reply by kickasskingy** [reply]
@Fox49 Any chance you can please post pictures of the exact soldering and wiring set-up you ended up with? (as well as what wires go to what corresponding wire on the spa pump).

Only I have the same model as you and have the same issues. I believe I have also fried my LLC also but have no way of testing it. Only way I can rule it out is if everything matches yours and it still doesn't work :/

I've also got mine used the resistors/caps, maybe this is the problem? did you need them?

Many thanks!

![WhatsApp Image 2024-07-30 at 20 02 59_f7368ae6](https://github.com/user-attachments/assets/5288bf89-1027-48a5-9c30-ad699f7aca41)

**Reply by Fox49** [reply]
Hi @kickasskingy 

I'm not using the second layout with capas and resistors, maybe I'll try later when I'll revert back to the default settings (maybe this winter when the spa will be stored).

From what I see on your picture and by comparing to mine :

-  CIO side : it seems you reverted TX and RX 
          - TX from CIO goes to D5 (cross-marked wire) -> Yellow one
          - RX from CIO goes to D2 (long line wire) -> White one
- DSP side : the wiring is wrong : on your setup (by deduction cause I don't see well on your photo) :
          - red is 5V
          - white is GND
          - Yellow is TX (goes to D3)
          - Black is RX (goes to D4)

Try this way, enable debug mode and give us the log.txt return

**Reply by kickasskingy** [reply]
Hi @Fox49 , 

Thanks for getting back to me.

It's been quite the ride hence the lag in responses! Basically, my ESP was faulty.... this was the first issue. It was restarting every 60s for no reason. Now this has been fixed we are closer as at least I can maintain a stable connection!

I have now wired it all in again, this time bypassing the caps/resistors. I still get 04 error for 10 minutes and then it kind of starts working. The 04 error passes and then I can control it remotely but not well. eg turn bubbles on and off. But cant adjust temp etc.

I ran a hwtest with your suggestions and here are the results:

Start test. Seq begins with HIGH, then alters.

Sending on D2, receiving on D4
-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0 // 50 errors out of 100
Sending on D5, receiving on D3
-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0 // 50 errors out of 100
Sending on D0, receiving on D0
1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1- // 50 errors out of 100
Sending on D4, receiving on D2
-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0 // 50 errors out of 100
Sending on D3, receiving on D5
-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0 // 50 errors out of 100
Sending on D0, receiving on D0
1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1- // 50 errors out of 100
End of test!
"1" or "0" indicates ERROR, depending on test state. "-" is good.
Switching cio pins 5s HIGH -> 5s LOW -> input
then DSP pins 5s HIGH -> 5s LOW -> input, repeating
Disconnect cables then reset chip when done!

Debug report:
2024-08-13 19:59:43UTC.  SW:2024-06-10-2016 
CIO-ESP:BC,3,17,0,1,1B,FD,0,0,0,0
DSP-ESP:BB,3,0,0,0,3,FD,0,0,0,0
ESP-CIO:BB,3,0,0,0,3,FD,0,0,0,0
ESP-DSP:BC,3,17,0,1,1B,FD,0,0,0,0
CIO msg count: 4 DSP msg count: 1072
CIO msg quality: 100.000000% DSP msg quality: 100.000000% (Only useful in 4 wire pumps)

Not really sure what to make of all this, but it doesn't look good! I haven't replaced the LLC - I don't have a spare atm.

I feel so close now!

![WhatsApp Image 2024-08-13 at 20 58 03_f75e1133](https://github.com/user-attachments/assets/8fa84525-0772-494b-b397-f15def0ed02c)

**Comment by piltous** [comment]
@Fox49 you answered yourself but I'm interested if it works when you go back to initial pin setup...

**Reply by Fox49** [reply]
I'm lazy to try it for now, but if/when I'll try, I'll let you know if it's still ok

**Comment by Fox49** [comment]
@visualapproach does the pump keeps its initial setup ? (I mean turn heat on automatically when temp. falls under the set point ? ) 

That being said I wonder if it automatically starts with the default setup... not sure..

**Reply by visualapproach** [reply]
I cannot guarantee that, but as long as the ESP is working and there is no bugs, there is antifreeze- and antiboil-functions. I have to restate the disclaimer though - I wont be responsible if it fails.
